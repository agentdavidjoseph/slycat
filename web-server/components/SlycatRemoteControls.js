import React, {Component} from 'react';
import client from "js/slycat-web-client";

export default class SlycatRemoteControls extends Component {
    constructor(props) {
      super(props);
      const display = this.populateDisplay();
      this.state = {
        remote_hosts: [],
        hostName: display.hostName?display.hostName:null,
        userName: display.userName?display.userName:null,
        session_exists: false,
        password: "",
        hostNames : []
      };
    }
    buildHostNames = (json) => {
      this.setState({hostNames:json});
    }
    checkRemoteStatus(hostName){
      console.log(hostName);
    }
    updateRemoteStatus(json){
      console.log(json);
    }
    getRemoteHosts = () => {
      client.get_configuration_remote_hosts_fetch(this.buildHostNames)
    }
    componentDidMount(){
      this.getRemoteHosts();
    }
    populateDisplay = () => {
      const display = {};
      if(localStorage.getItem("slycat-remote-controls-hostname")){
        display.hostName = localStorage.getItem("slycat-remote-controls-hostname") ?
        localStorage.getItem("slycat-remote-controls-hostname"):null;
      };
      if(localStorage.getItem("slycat-remote-controls-username")){
        display.userName = localStorage.getItem("slycat-remote-controls-username") ?
        localStorage.getItem("slycat-remote-controls-username"):null;
      };
      return display;
    }
    onValueChange = (value, type) => {
      switch(type) {
        case "userName":
          localStorage.setItem("slycat-remote-controls-username", value);
          this.setState({userName: value})
          break;
        case "hostName":
          localStorage.setItem("slycat-remote-controls-hostname", value);
          this.checkRemoteStatus(value)
          this.setState({hostName: value})
          break;
        case "password":
          this.setState({password: value})
          break;
        default:
          throw new Error("bad Case");
      }

    }
    componentWillUnmount = () => {
      const display = this.populateDisplay();
      const state = {
        remote_hosts: [],
        enable: true,
        hostName: display.hostName?display.hostName:null,
        userName: display.userName?display.userName:null,
        use_remote: () =>{},
        session_exists: false,
        password: null
      };
      this.setState(state);
    }

    render() {
      console.log(`state ${JSON.stringify(this.state)}`)
      const hostNamesJSX = this.state.hostNames.map((hostnameObject, i) => {
        return (
        <a className="dropdown-item" key={i} onClick={(e)=>this.onValueChange(e.target.text, "hostName")}>{hostnameObject.hostname}</a>
        )
      })
      return (
        <div>
          <div className="form-group row">
            <label className="col-sm-2 col-form-label">Hostname</label>
            <div className="col-sm-10">
              <div className="input-group">
                <div className="dropdown">
                  <button className="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                  </button>
                  <div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
                  {hostNamesJSX}
                  </div>
                </div>
                <input className="form-control" value={this.state.hostName?this.state.hostName:""} type="text" 
                onChange={(e)=>this.onValueChange(e.target.value, "hostName")}></input>
              </div>
            </div>
          </div>
          {/* <!-- ko if: ispasswordrequired.ssh_passwordrequired --> */}
          {true?<div>
          <div className="form-group row" data-bind-old="visible: !session_exists()">
            <label className="col-sm-2 col-form-label">Username</label>
            <div className="col-sm-10">
              <input className="form-control" type="text" 
              value={this.state.userName?this.state.userName:""} 
              onChange={(e)=>this.onValueChange(e.target.value, "userName")}></input>
            </div>
          </div>
          <div className="form-group row" data-bind-old="visible: !session_exists()">
            <label className="col-sm-2 col-form-label">Password</label>
            <div className="col-sm-10">
              <input className="form-control" type="password" 
              onChange={(e)=>this.onValueChange(e.target.value, "password")}></input>
            </div>
          </div>
          </div>:null}
          <div className="row">
            <div className="col-sm-offset-2 col-sm-10">
              <div className="alert fade" role="alert"></div>
            </div>
          </div>

        </div>
      );
    };
}