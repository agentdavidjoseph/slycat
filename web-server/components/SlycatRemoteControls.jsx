import React, {Component} from 'react';
import client from "../js/slycat-web-client";

export default class SlycatRemoteControls extends Component {
    constructor(props) {
      super(props);
      const display = this.populateDisplay();
      this.state = {
        remote_hosts: [],
        hostName: display.hostName?display.hostName:null,
        userName: display.userName?display.userName:null,
        session_exists: null,
        password: "",
        hostNames : [],
        loaded: false
      };
    }

    checkRemoteStatus = (hostName) => {
      return client.get_remotes_fetch(hostName)
        .then((json) => {
          this.setState({session_exists:json.status,loaded:true});
          console.log(json);
      }).then(() => {this.props.callBack(this.state.hostName, this.state.userName,
          this.state.password, this.state.session_exists,);
      });
    };

    getRemoteHosts = () => {
      return client.get_configuration_remote_hosts_fetch()
        .then((json)=>{
          this.setState({hostNames:json});
        })
    };

    async componentDidMount(){
      await this.checkRemoteStatus(this.state.hostName);
      await this.getRemoteHosts();
    }

    populateDisplay = () => {
      const display = {};
      if(localStorage.getItem("slycat-remote-controls-hostname")){
        display.hostName = localStorage.getItem("slycat-remote-controls-hostname") ?
        localStorage.getItem("slycat-remote-controls-hostname"):null;
      }
      if(localStorage.getItem("slycat-remote-controls-username")){
        display.userName = localStorage.getItem("slycat-remote-controls-username") ?
        localStorage.getItem("slycat-remote-controls-username"):null;
      }
      return display;
    };

    onValueChange = (value, type) => {
      switch(type) {
        case "userName":
          localStorage.setItem("slycat-remote-controls-username", value);
          this.setState({userName: value});
          break;
        case "hostName":
          localStorage.setItem("slycat-remote-controls-hostname", value);
          this.checkRemoteStatus(value);
          this.setState({hostName: value});
          break;
        case "password":
          this.setState({password: value});
          break;
        default:
          throw new Error("bad Case");
      }
    };

    componentWillUnmount() {
      const display = this.populateDisplay();
      const state = {
        remote_hosts: [],
        enable: true,
        hostName: display.hostName?display.hostName:null,
        userName: display.userName?display.userName:null,
        session_exists: false,
        password: null,
        loaded: false
      };
      this.setState(state);
    }

    connect = async () => {
      client.post_remotes_fetch({
        parameters: {
          hostname: this.state.hostName,
          username: this.state.userName,
          password: this.state.password,
        }
      }).then(() => {
        this.checkRemoteStatus(this.state.hostName);
        console.log("Remote session created.");
      });
    };

    render() {
      //make sure our data is loaded before we render
      if(!this.state.loaded){
        return (<div></div>)
      }
      const hostNamesJSX = this.state.hostNames.map((hostnameObject, i) => {
        return (
        <a className="dropdown-item" key={i} onClick={(e)=>this.onValueChange(e.target.text, "hostName")}>{hostnameObject.hostname}</a>
        )
      })
      return (
        <div>
          <div className="form-group row">
            <label className="col-sm-3 col-form-label">Hostname</label>
            <div className="col-sm-10">
              <div className="input-group">
                <div className="dropdown">
                  <button className="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                  </button>
                  <div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
                  {hostNamesJSX}
                  </div>
                </div>
                <input className="form-control" value={this.state.hostName?this.state.hostName:""} type="text" 
                onChange={(e)=>this.onValueChange(e.target.value, "hostName")}></input>
              </div>
            </div>
          </div>

          {!this.state.session_exists?
          <div>
            <div className="form-group row">
              <label className="col-sm-3 col-form-label">Username</label>
              <div className="col-sm-10">
                <input className="form-control" type="text"
                value={this.state.userName?this.state.userName:""}
                onChange={(e)=>this.onValueChange(e.target.value, "userName")}></input>
              </div>
            </div>
            <div className="form-group row" data-bind-old="visible: !session_exists()">
              <label className="col-sm-3 col-form-label">Password</label>
              <div className="col-sm-10">
                <input className="form-control" type="password"
                onChange={(e)=>this.onValueChange(e.target.value, "password")}></input>
              </div>
            </div>
            <div className="form-group row" >
              <div className="col-sm-10" style={{marginTop: '15px'}}>
                <button type='button' className='btn btn-primary' onClick={this.connect}>
                  Connect
                </button>
              </div>
            </div>
          </div>:
          null}


        </div>
      );
    };
}