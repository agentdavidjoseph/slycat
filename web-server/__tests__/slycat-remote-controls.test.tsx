import SlycatRemoteControls from "../components/SlycatRemoteControls";
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe("when we load the SlycatRemoteControls", () =>{

  let component: renderer.ReactTestRenderer;
  let DefaultInstance: any;
  beforeEach(async () => {
    global.fetch.mockReset();
    global.fetch.mockResponse(JSON.stringify({ data: '12345' }))
    component = await renderer.create(
      <SlycatRemoteControls callBack={()=>{}}/>
    );
    DefaultInstance = component.getInstance();
  });

  test('we can render without crashing', async () => {
    expect(await shallow(<SlycatRemoteControls userName={"test"}/>)).toBeTruthy();
  });

  test('we can render the component', () => {
    expect(component).toBeTruthy();
  });

  test('we have expected state on initial load', () => {
    
    expect(DefaultInstance.props.userName).toBe("test");
  });

  test('we expect the correct props to populate', () => {
    expect(DefaultInstance.state.path).toBe("/");
  });

  test('we expect the correct props to populate', () => {
    expect(DefaultInstance.render()).toMatchSnapshot();
  });
  
});
