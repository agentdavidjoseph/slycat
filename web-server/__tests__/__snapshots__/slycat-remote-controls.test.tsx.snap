// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when we load the SlycatRemoteControls we expect the correct props to populate 1`] = `
<div>
  <div
    className="form-group row"
  >
    <label
      className="col-sm-3 col-form-label"
    >
      Hostname
    </label>
    <div
      className="col-sm-10"
    >
      <div
        className="input-group"
      >
        <div
          className="dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            className="btn btn-secondary dropdown-toggle"
            data-toggle="dropdown"
            id="dropdownMenuButton"
            type="button"
          />
          <div
            aria-labelledby="dropdownMenuButton"
            className="dropdown-menu"
          />
        </div>
        <input
          className="form-control"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div>
    <div
      className="form-group row"
    >
      <label
        className="col-sm-3 col-form-label"
      >
        Username
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group row"
      data-bind-old="visible: !session_exists()"
    >
      <label
        className="col-sm-3 col-form-label"
      >
        Password
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          onChange={[Function]}
          type="password"
        />
      </div>
    </div>
    <div
      className="form-group row"
    >
      <div
        className="col-sm-10"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Connect
        </button>
      </div>
    </div>
  </div>
</div>
`;
